docs/Using-SRIOV-and-Clear-Containers.md

```diff
@@ -0,0 +1,281 @@
+
+See the following steps one through three if you need to rebuild the kernel.
+
+The following steps update the SR-IOV host system. The steps are based on 
+Ubuntu 16.04 distribution. If you use a different distribution, make 
+appropriate adjustments to the commands.
+
+1. Grab kernel sources:
+
+   ```
+   $ sudo apt-get install linux-source-4.10.0
+   $ sudo apt-get install linux-headers-4.10.0
+   $ cd /usr/src/linux-source-4.10.0/
+   $ sudo tar -xvf linux-source-4.10.0.tar.bz2
+   $ cd linux-source-4.10.0
+   $ sudo apt-get install libssl-dev
+   ```
+
+2. Examine and update the `config` file if necessary: 
+
+   ```
+   $ sudo cp /boot/config-4.8.0-36-generic .config
+   $ # verify resulting .config does not have NOIOMMU set; ie: `CONFIG_VFIO_NOIOMMU` is not set
+   $ # verify `CONFIG_VFIO_IOMMU_TYPE1`, `CONFIG_VFIO=m` and `CONFIG_VFIO_PCI=m` are set as well.
+   $ sudo make olddefconfig 
+   ```
+
+   You might want to modify the kernel `Makefile` to add a unique identifier 
+   to the `EXTRAVERSION` variable prior to running the make. Including the `EXTRAVERSION` 
+   variable causes the `uname -r` command to indicate that a customized kernel is
+   installed and running.
+
+3. Build and install the kernel:
+
+   ```
+   $ make -j3 #and go get coffee
+   $ make modules -j 3
+   $ sudo make modules_install
+   $ sudo make install
+   ```
+
+4. Edit grub to enable `intel-iommu`:
+
+   ```
+   edit /etc/default/grub and add intel_iommu=on to cmdline:
+   - GRUB_CMDLINE_LINUX=""
+   + GRUB_CMDLINE_LINUX="intel_iommu=on"
+   sudo update-grub
+   ```
+
+5. Reboot the system and verify:
+
+   Host system should be ready now. Reboot the system.
+   ```
+   sudo reboot
+   ```
+
+   To verify the kernel version and the kernel command line, take a look at
+   `/proc/version` and `/proc/cmdline`
+
+6. Verify Intel VT-d has been initialized:
+   
+   To check if Intel VT-d initialized correctly, look for the following 
+   line in the `dmesg` output:
+   ```
+   DMAR: Intel(R) Virtualization Technology for Directed I/O
+   ```
+
+   Older kernels use a different prefix (e.g. PCI-DMA)
+   ```
+   PCI-DMA: Intel(R) Virtualization Technology for Directed I/O
+   ```
+
+7. Add the `vfio-pci` module
+
+   ```
+   sudo modprobe vfio-pci
+   ```
+
+8. If necessary, add PCI quirk for SR-IOV NIC
+
+   ```
+   $ find /sys/kernel/iommu_groups/ -type l
+   ```
+   The previous command verifies that your NIC appears in its own IOMMU group
+   and no other devices appear in the same group. In the rare case where your 
+   PCI NIC does not appear in its own group, the NIC might not support ACS or 
+   you may have chosen to build and run an old kernel. Depending on your NIC 
+   and if it indeed enforces isolation, you might resolve this by adding a 
+   `pcie_acs_override=` option to your kernel command line and reboot. 
+   You can find detailed information about this option at [PCIE-ACS-override-option](https://lkml.org/lkml/2013/5/30/513).
```

Comment:
See [PCIE-ACS-override-option](https://lkml.org/lkml/2013/5/30/513) for detailed information about this option.